library/src/scala/annotation/MainAnnotation.scala

```diff
@@ -0,0 +1,43 @@
+package scala.annotation
+
+/** MainAnnotation provides the functionality for a compiler-generated main class.
+ *  It links a compiler-generated main method (call it compiler-main) to a user
+ *  written main method (user-main).
+ *  The protocol of calls from compiler-main is as follows:
+ *
+ *    - create a `command` with the command line arguments,
+ *    - for each parameter of user-main, a call to `command.argGetter`,
+ *      or `command.argsGetter` if is a final varargs parameter,
+ *    - a call to `command.run` with the closure of user-main applied to all arguments.
+ */
+trait MainAnnotation extends StaticAnnotation:
+
+  /** The class used for argument string parsing. E.g. `scala.util.CommandLineParser.FromString`,
+   *  but could be something else
+   */
+  type ArgumentParser[T]
+
+  /** The required result type of the main function */
+  type MainResultType
+
+  /** A new command with arguments from `args` */
+  def command(args: Array[String], commandName: String, docComment: String): Command
+
+  /** A class representing a command to run */
+  abstract class Command(val commandName: String, val docComment: String):
```

Comment:
Now that we have `def command` we do not need the arguments for `Command`
```suggestion
  trait Command:
```